Mage/src/main/java/mage/abilities/effects/common/GainsChoiceOfAbilitiesEffect.java

```diff
@@ -0,0 +1,120 @@
+
+package mage.abilities.effects.common;
+
+import mage.abilities.Ability;
+import mage.abilities.Mode;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.continuous.GainAbilityTargetEffect;
+import mage.choices.Choice;
+import mage.choices.ChoiceImpl;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.targetpointer.FixedTargets;
+
+import java.util.*;
+
+/**
+ * Something gains one ability of the player's choice from a list of abilities until end of turn.
+ * Affects any targets set and optionally the source as well. Have no targets to affect only the source.
+ *
+ * @author notgreat, TheElk801
+ */
+public class GainsChoiceOfAbilitiesEffect extends OneShotEffect {
```

Comment:
Must be named as xxxTargetEffect

Reply:
This effect is being used both for "~ gains your choice" and for "target gains your choice" effects (and one with both at once). Should it really be named TargetEffect when half the uses of it don't target?

Reply:
If target setup required then it must be xxxTargetEffect. If it works without target (e.g. for source only) then no needs in xxxTargetEffect name.